platform/posix/transport/src/mbedtls_pkcs11_posix.c

```diff
@@ -0,0 +1,958 @@
+{
+    CK_RV ret = CKR_OK;
+    CK_ATTRIBUTE template[ 2 ] = { 0 };
+    mbedtls_pk_type_t keyAlgo = ( mbedtls_pk_type_t ) ~0;
+
+    /* Get the handle of the device private key. */
+    ret = xFindObjectWithLabelAndClass( pCtx->p11Session,
+                                        pPrivateKeyLabel,
+                                        strlen( pPrivateKeyLabel ),
+                                        CKO_PRIVATE_KEY,
+                                        &pCtx->p11PrivateKey );
+
+    if( ( ret == CKR_OK ) && ( pCtx->p11PrivateKey == CK_INVALID_HANDLE ) )
+    {
+        ret = CK_INVALID_HANDLE;
+        LogError( ( "Could not find private key." ) );
+    }
+
+    /* Query the device private key type. */
+    if( ret == CKR_OK )
+    {
+        template[ 0 ].type = CKA_KEY_TYPE;
+        template[ 0 ].pValue = &pCtx->keyType;
+        template[ 0 ].ulValueLen = sizeof( CK_KEY_TYPE );
+        ret = pCtx->pP11FunctionList->C_GetAttributeValue( pCtx->p11Session,
+                                                           pCtx->p11PrivateKey,
+                                                           template,
+                                                           1 );
+    }
+
+    /* Map the PKCS #11 key type to an mbedTLS algorithm. */
+    if( ret == CKR_OK )
+    {
+        switch( pCtx->keyType )
+        {
+            case CKK_RSA:
+                keyAlgo = MBEDTLS_PK_RSA;
+                break;
+
+            case CKK_EC:
+                keyAlgo = MBEDTLS_PK_ECKEY;
+                break;
+
+            default:
+                ret = CKR_ATTRIBUTE_VALUE_INVALID;
+                break;
+        }
+    }
+
+    /* Map the mbedTLS algorithm to its internal metadata. */
+    if( ret == CKR_OK )
+    {
+        memcpy( &pCtx->privKeyInfo, mbedtls_pk_info_from_type( keyAlgo ), sizeof( mbedtls_pk_info_t ) );
+
+        pCtx->privKeyInfo.sign_func = privateKeySigningCallback;
+        pCtx->privKey.pk_info = &pCtx->privKeyInfo;
+        pCtx->privKey.pk_ctx = pCtx;
+    }
+
+    return ret;
+}
+
+/*-----------------------------------------------------------*/
+
+static int32_t privateKeySigningCallback( void * pContext,
+                                          mbedtls_md_type_t mdAlg,
+                                          const unsigned char * pHash,
+                                          size_t hashLen,
+                                          unsigned char * pSig,
+                                          size_t * pSigLen,
+                                          int32_t ( *pRng )( void *,
+                                                             unsigned char *,
+                                                             size_t ),
+                                          void * pRngContext )
+{
+    CK_RV ret = CKR_OK;
+    int32_t result = 0;
+    MbedtlsPkcs11Context_t * pMbedtlsPkcs11Context = ( MbedtlsPkcs11Context_t * ) pContext;
+    CK_MECHANISM mech = { 0 };
+    CK_BYTE toBeSigned[ 256 ];
```

Comment:
Why is 256 bytes chosen as the hash len? Is this the standard hash size used across all TLS handshakes for verifying certificate?

Reply:
That is just a buffer size big enough to hold the types of hashes.